File: 332. Reconstruct Itinerary.py
from typing import List
from collections import defaultdict

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        mp = defaultdict(list)
        cnt = dict()
        for ticket in tickets:
            mp[ticket[0]].append(ticket[1])
        for i in mp:
            mp[i].sort()
            cnt[i] = 0
        ans = []

        def dfs(cur: str) -> None:
            nonlocal cnt, ans
            n = len(mp[cur])
            while cur in cnt and cnt[cur] < n:
                cnt[cur] += 1
                dfs(mp[cur][cnt[cur] - 1])
            ans.append(cur)

        dfs("JFK")
        return ans[::-1]

File: 332. Reconstruct Itinerary.cpp
#include <iostream>
#include <algorithm>
#include <set>
#include <map>
using namespace std;

class Solution
{
public:
    vector<string> findItinerary(vector<vector<string>>& tickets)
    {
        map<string,multiset<string>> mp;
        for(auto& ticket:tickets)
        {
            mp[ticket[0]].insert(ticket[1]);
        }
        vector<string> ans;
        
        function<void(string)> dfs=[&](string p)
        {
            while(!mp[p].empty())
            {
                string next=*mp[p].begin();
                mp[p].erase(mp[p].begin());
                dfs(next);
            }
            ans.push_back(p);
        };
        
        dfs("JFK");
        reverse(ans.begin(),ans.end());
        return ans;
    }
};
